extensions/rich_text_components/rte-output-display.component.ts

```diff
@@ -0,0 +1,178 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Component for the showing rich text.
+ */
+
+import { TemplatePortal } from '@angular/cdk/portal';
+import { AfterViewInit, ChangeDetectorRef, Component, Directive, ElementRef, Input, SimpleChanges, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
+import { downgradeComponent } from '@angular/upgrade/static';
+import { OppiaRteParserService, OppiaRteNode, TextNode } from 'services/oppia-rte-parser.service';
+
+type PortalTree = (TemplatePortal<unknown> | PortalTree) [];
+
+@Component({
+  selector: 'oppia-rte-output-display',
+  templateUrl: './rte-output-display.component.html',
+  styleUrls: []
+})
+export class RteOutputDisplayComponent implements AfterViewInit {
+  // Native HTML elements.
+  @ViewChild('p') pTagPortal: TemplateRef<unknown>;
+  @ViewChild('span') spanTagPortal: TemplateRef<unknown>;
+  @ViewChild('ol') olTagPortal: TemplateRef<unknown>;
+  @ViewChild('li') liTagPortal: TemplateRef<unknown>;
+  @ViewChild('ul') ulTagPortal: TemplateRef<unknown>;
+  @ViewChild('pre') preTagPortal: TemplateRef<unknown>;
+  @ViewChild('strong') strongTagPortal: TemplateRef<unknown>;
+  @ViewChild('blockquote') blockquoteTagPortal: TemplateRef<unknown>;
+  @ViewChild('em') emTagPortal: TemplateRef<unknown>;
+  @ViewChild('text') textTagPortal: TemplateRef<unknown>;
+  // Oppia Non interactive.
+  @ViewChild('collapsible') collapsibleTagPortal: TemplateRef<unknown>;
+  @ViewChild('image') imageTagPortal: TemplateRef<unknown>;
+  @ViewChild('link') linkTagPortal: TemplateRef<unknown>;
+  @ViewChild('math') mathTagPortal: TemplateRef<unknown>;
+  @ViewChild('skillreview') skillreviewTagPortal: TemplateRef<unknown>;
+  @ViewChild('svgdiagram') svgdiagramTagPortal: TemplateRef<unknown>;
+  @ViewChild('tabs') tabsTagPortal: TemplateRef<unknown>;
+  @ViewChild('video') videoTagPortal: TemplateRef<unknown>;
+  @Input() rteString: string;
+  node: OppiaRteNode | string = '';
+  show = false;
+  portalTree: PortalTree = [];
+
+  constructor(
+    private _viewContainerRef: ViewContainerRef,
+    private cdRef: ChangeDetectorRef,
+    private elementRef: ElementRef,
+    private oppiaHtmlParserService: OppiaRteParserService
+  ) {}
+
+  private _updateNode(): void {
+    if (this.rteString === undefined || this.rteString === null) {
+      return;
+    }
+    // When there are trailing spaces in the HTML, CKEditor adds &nbsp;
+    // to the HTML (eg: '<p> Text &nbsp; &nbsp; %nbsp;</p>'), which can
+    // lead to UI issues when displaying it. Hence, the following block
```

Comment:
Can you be more specific about what the "UI issues" are?

Reply:
I have no idea here but my guess is that ckeditor adds a bunch of nbsp; and that doesn't cause a problem in the ck-editor view but cause problem when we show it. I took these from angular-html-bind (https://github.com/oppia/oppia/blob/develop/core/templates/directives/angular-html-bind.directive.ts#L37-L51).